### PR TITLE
Remove dependence on tid; improve docs and typing for BioBERT

### PR DESCRIPTION
**Pull request by mingkaid**
This PR fixes removes the dependence of `BioBERTNERPredictor` on the sequential order of `tid`s in a `DataPack` object. 

### Description of changes
1. The routine reconstructing name entities in `BioBERTNERPredictor` class is tweaked to not rely on the sequential order of `tid`s in a `DataPack` object
2. The code is refactored to use clearer terminologies
3. Documentation is added to explain the routines 

### Test Conducted
Tested on several short sentences from CORD-19. Tests on MIMIC3 notes are still pending

### CODE REVIEW

File: examples/biobert_ner/bio_ner_predictor.py [hunk 1]
```diff
@@ -105,58 +105,83 @@ def predict(self, data_batch: Dict[str, Dict[str, List[str]]]) \
...
-            print('Encounter get entry error.')
+    def _complete_entity(self,
+                         subword_entities: List[Dict[str, Any]],
+                         data_pack: DataPack,
+                         tids: List[int]) -> Tuple[int, int]:
+        """ Complete entity span from predicted subword entities
 
-        return first_tid, last_tid
+        Start from the first subword with predicted entity. If this entity
+        is a subword (e.g. "##on"), then move on to the previous subword until
```
Comment (hunterhector) [inline examples/biobert_ner/bio_ner_predictor.py:115]: It looks like the semantic of `is_subword` is actually `is_not_first_segment`. Maybe we can rename this attribute to make it more intuitive? Or we can use `is_first_segment` with reversed boolean values?

File: examples/biobert_ner/bio_ner_predictor.py [hunk 2]
```diff
@@ -105,58 +105,83 @@ def predict(self, data_batch: Dict[str, Dict[str, List[str]]]) \
...
+
+        last_idx: int = subword_entities[-1]['idx']
+        while last_idx < len(tids) - 1 and data_pack.get_entry(
+            tids[last_idx + 1]).is_subword:
+            last_idx += 1
+
+        return first_idx, last_idx
+
+    def _compose_entities(self,
+                          entities: List[Dict[str, Any]],
```
Comment (hunterhector) [inline examples/biobert_ner/bio_ner_predictor.py:134]: Could you document the expected input args and return types?

File: examples/biobert_ner/bio_ner_predictor.py [hunk 3]
```diff
@@ -165,32 +190,36 @@ def pack(self, data_pack: DataPack,
         ner to the original subwords and convert predictions to something that
         makes sense in a word-by-word segmentation
         """
-
+        # print(list(data_pack.get(Sentence)))
```
Comment (hunterhector) [inline examples/biobert_ner/bio_ner_predictor.py:193]: We can remove this line.

File: examples/biobert_ner/bio_ner_predictor.py [hunk 4]
```diff
@@ -165,32 +190,36 @@ def pack(self, data_pack: DataPack,
...
 
-            entity_groups = self._group_entities(entities, data_pack, tids)
+            entity_groups = self._compose_entities(entities, data_pack, tids)
             # Add NER tags and create EntityMention ontologies.
-            for first_tid, last_tid in entity_groups:
+            for first_idx, last_idx in entity_groups:
                 first_token: Subword = data_pack.get_entry(  # type: ignore
-                    first_tid)
+                    tids[first_idx])
                 first_token.ner = 'B-' + self.ft_configs.ner_type
```
Comment (hunterhector) [inline examples/biobert_ner/bio_ner_predictor.py:212]: We can make it clear in the readme why we only have one ner type in this example.

File: examples/biobert_ner/bio_ner_predictor.py [hunk 5]
```diff
@@ -165,32 +190,36 @@ def pack(self, data_pack: DataPack,
...
+                begin = first_token.span.begin
 
-                for tid in range(first_tid + 1, last_tid + 1):
+                for tid in tids[(first_idx + 1):(last_idx + 1)]:
                     token: Subword = data_pack.get_entry(tid)  # type: ignore
                     token.ner = 'I-' + self.ft_configs.ner_type
```
Comment (hunterhector) [inline examples/biobert_ner/bio_ner_predictor.py:217]: It is actually not important to add `ner` to the tokens. We can actually save some space by not setting them.

File: examples/biobert_ner/bio_ner_predictor.py [hunk 6]
```diff
@@ -105,58 +105,83 @@ def predict(self, data_batch: Dict[str, Dict[str, List[str]]]) \
...
-        if entity_groups_disagg:
-            entity_groups_disagg = self._fill_sub_entities(entity_groups_disagg,
-                                                           data_pack,
-                                                           tids)
-            entity_groups.append(entity_groups_disagg)
+        if subword_entities:
+            subword_entities = self._complete_entity(subword_entities,
+                                                     data_pack,
+                                                     tids)
+            complete_entities.append(subword_entities)
```
Comment (hunterhector) [inline examples/biobert_ner/bio_ner_predictor.py:182]: So the `_complete_entity` function will try to include the whole word from the subword annotations. Is it possible that after these auto-expansion, we will have overlapping entities somehow?

**Review by hunterhector — CHANGES_REQUESTED** [review]
I think this is much clear with the documentation. Could you resolve some small comments and also make pass mypy on travis?

File: examples/biobert_ner/bio_ner_predictor.py (hunk 1 above)
Comment (mingkaid) [reply examples/biobert_ner/bio_ner_predictor.py:115]: Makes sense. I can change the name

File: examples/biobert_ner/bio_ner_predictor.py (hunk 2 above)
Comment (mingkaid) [reply examples/biobert_ner/bio_ner_predictor.py:134]: All expected input args and return types are documented. Are you referring to any specific function signature?

File: examples/biobert_ner/bio_ner_predictor.py (hunk 6 above)
Comment (mingkaid) [reply examples/biobert_ner/bio_ner_predictor.py:182]: No it won't. `_complete_entity` will include `Subword` annotations until the first `Subword` that is a first segment, and the last `Subword` that's not a first segment. Whereas we always flush the `subword_entities` list when we see a "B" annotation that *is* a first segment. This means we'll never cross the line.

File: examples/biobert_ner/bio_ner_predictor.py (hunk 4 above)
Comment (mingkaid) [reply examples/biobert_ner/bio_ner_predictor.py:212]: OK

File: examples/biobert_ner/bio_ner_predictor.py (hunk 5 above)
Comment (mingkaid) [reply examples/biobert_ner/bio_ner_predictor.py:217]: Sounds good

Comment (mingkaid) [reply examples/biobert_ner/bio_ner_predictor.py:217]: Sounds good

File: examples/biobert_ner/bio_ner_predictor.py (hunk 3 above)
Comment (mingkaid) [reply examples/biobert_ner/bio_ner_predictor.py:193]: I will

File: .travis.yml [hunk 7]
```diff
@@ -14,6 +14,7 @@ install:
   - pip install --upgrade pip
   - pip install --progress-bar off torch==$TORCH_VER
   - pip install --progress-bar off tensorflow==$TENSORFLOW_VER
+  - pip install --progress-bar off spacy==2.2.4
```
Comment (hunterhector) [inline .travis.yml:17]: What cause this strict dependency to spacy?

Comment (hunterhector) [reply .travis.yml:17]: It seems that AllenNLP may change the dependencies on their 1.0.0 version on Pip. Not sure what can cause this. Someone is working to fix this here: https://github.com/asyml/forte/pull/321

**Comment by hunterhector** [comment]
Hi, I have fixed the traivis and mypy issues in master, could you resolve the conflicts and merge the master?

**Review by hunterhector — APPROVED** [review]

**Pull request merged**